File: backends/Auther/Services/Auth/TokenService.cs
using Microsoft.EntityFrameworkCore;

using Auther.Data;
using Auther.Interfaces.Auth;
using Auther.Models;
using System.Security.Cryptography;

namespace Auther.Services.Auth
{

    public class TokenService : ITokenRepository
    {
        private readonly AppDbContext _context;
        private readonly IMailService _emailService;
        private readonly ILogger<TokenService> _logger;

        public TokenService(AppDbContext context, 
                            IMailService emailService, 
                            ILogger<TokenService> logger
        )
        {
            _context = context;
            _emailService = emailService;
            _logger = logger;
        }
        private static string GenerateRandomCode()
        {
            string Characters = "0123456789";
            int CodeLength = 6;
            byte[] randomBytes = new byte[1];
            string code = string.Empty;

            for (int i = 0; i < CodeLength; i++)
            {
                RandomNumberGenerator.Fill(randomBytes);
                int randomIndex = randomBytes[0] % Characters.Length;
                code += Characters[randomIndex];
            }

            return code;
        }


        public async Task<VerificationToken?> GetVerificationTokenByTokenAsync(string token)
        {
            var existingToken = await _context.VerificationTokens
                    .FirstOrDefaultAsync(t => t.Token == token);

            return existingToken;
        }

        public async Task<VerificationToken?> GetVerificationTokenByEmailAsync(string email)
        {
            var existingToken = await _context.VerificationTokens
                    .FirstOrDefaultAsync(t => t.Email == email);

            return existingToken;
        }


        public async Task<TwoFactorToken?> GetTwoFactorTokenByTokenAsync(string token)
        {
            var existingToken = await _context.TwoFactorTokens
                    .FirstOrDefaultAsync(t => t.Token == token);

            return existingToken;
        }

        public async Task<TwoFactorToken?> GetTwoFactorTokenByEmailAsync(string email)
        {
            var existingToken = await _context.TwoFactorTokens
                    .FirstOrDefaultAsync(t => t.Email == email);

            return existingToken;
        }




        public async Task<PasswordResetToken?> GetPasswordResetTokenByTokenAsync(string token)
        {
            var existingToken = await _context.PasswordResetTokens
                    .FirstOrDefaultAsync(t => t.Token == token);

            return existingToken;
        }

        public async Task<PasswordResetToken?> GetPasswordResetTokenByEmailAsync(string email)
        {
            var existingToken = await _context.PasswordResetTokens
                    .FirstOrDefaultAsync(t => t.Email == email);

            return existingToken;
        }




        //Deletes existing tokens

        private async Task DeleteExistingTokenAsync(string email, string tokenType)
        {
            if (tokenType == "verificationToken")
            {
                var existingToken = await _context.VerificationTokens
                    .FirstOrDefaultAsync(t => t.Email == email);

                if (existingToken != null)
                {
                    _context.VerificationTokens.Remove(existingToken);
                    await _context.SaveChangesAsync();
                }
            }

            if (tokenType == "twoFactor")
            {
                var existingToken = await _context.TwoFactorTokens
                    .FirstOrDefaultAsync(t => t.Email == email);

                if (existingToken != null)
                {
                    _context.TwoFactorTokens.Remove(existingToken);
                    await _context.SaveChangesAsync();
                }
            }

            if (tokenType == "passwordReset")
            {
                var existingToken = await _context.PasswordResetTokens
                    .FirstOrDefaultAsync(t => t.Email == email);

                if (existingToken != null)
                {
                    _context.PasswordResetTokens.Remove(existingToken);
                    await _context.SaveChangesAsync();
                }
            }
        }



        public async Task<VerificationToken> GenerateVerificationTokenAsync(string email)
        {
            string token = GenerateRandomCode();
            DateTime expires = DateTime.UtcNow.AddMinutes(10);

            try
            {
                await DeleteExistingTokenAsync(email, "verificationToken");

                var verificationToken = new VerificationToken
                {
                    Email = email,
                    Token = token,
                    Expires = expires
                };

                if (_context.VerificationTokens != null)
                {
                    await _context.VerificationTokens.AddAsync(verificationToken);
                    await _context.SaveChangesAsync();
                }
                else
                {
                    throw new InvalidOperationException("VerificationTokens DbSet is not initialized.");
                }

                await _emailService.SendVerificationEmailAsync(email, token);
                return verificationToken;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error generating verification token for {Email}", email);
                throw new Exception("Failed to generate verification token", ex);
            }
        }


        public async Task<TwoFactorToken> GenerateTwoFactorTokenAsync(string email)
        {
            string token = GenerateRandomCode();
            DateTime expires = DateTime.UtcNow.AddMinutes(10);

            try
            {
                await DeleteExistingTokenAsync(email, "twoFactor");

                var twoFactorToken = new TwoFactorToken
                {
                    Email = email,
                    Token = token,
                    Expires = expires
                };

                if (_context.TwoFactorTokens != null)
                {
                    await _context.TwoFactorTokens.AddAsync(twoFactorToken);
                    await _context.SaveChangesAsync();
                }
                else
                {
                    throw new InvalidOperationException("VerificationTokens DbSet is not initialized.");
                }
                return twoFactorToken;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error generating two factor token for {Email}", email);
                throw new Exception("Failed to generate two factor token", ex);
            }
        }




        public async Task<PasswordResetToken> GeneratePasswordResetTokenAsync(string email)
        {
            string token = Guid.NewGuid().ToString();
            DateTime expires = DateTime.UtcNow.AddMinutes(10);

            try
            {
                await DeleteExistingTokenAsync(email, "passwordReset");

                var passwordResetToken = new PasswordResetToken
                {
                    Email = email,
                    Token = token,
                    Expires = expires
                };

                if (_context.TwoFactorTokens != null)
                {
                    await _context.PasswordResetTokens.AddAsync(passwordResetToken);
                    await _context.SaveChangesAsync();
                }
                else
                {
                    throw new InvalidOperationException("PasswordResetTokens DbSet is not initialized.");
                }
                return passwordResetToken;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error generating password reset token for {Email}", email);
                throw new Exception("Failed to generate password reset token", ex);
            }
        }

    }
}

File: backends/fastapi/app/api/services/token/token_service.py
from typing import Any
from pydantic import EmailStr
import random
from datetime import datetime, timedelta
import pytz

from app.prisma import prisma
from app.api.services.email.email_service import EmailService
from app.api.services.token.i_token_service import ITokenService

class TokenService(ITokenService):
    def __init__(self, email_service: EmailService):
        self.prisma = prisma
        self.email_service = email_service
    
    def generate_random_code(self, length: int = 6) -> str:
        characters = '0123456789'
        code = ''.join(random.choice(characters) for _ in range(length))
        return code
    
    
    
    
    
    
    async def get_verification_token_by_email(self, email: str) -> Any:
        token = await prisma.verificationtoken.find_first(
            where={"email": email}
        )
        return token
    
    
    async def get_verification_token_by_token(self, token: str) -> Any:
        token_record = await prisma.verificationtoken.find_unique(
            where={"token": token}
        )
        return token_record
    
    
    
    
    async def get_two_factor_token_by_token(self, token: str) -> Any:
        token_record = await prisma.twofactortoken.find_unique(
            where={"token": token}
        )
        return token_record
    
    
    async def get_two_factor_token_by_email(self, email: str) -> Any:
        token = await prisma.twofactortoken.find_first(
            where={"email": email}
        )
        return token
    
    
    
    
    
    
    async def delete_existing_token(self, email: str, token_type: str) -> None:
        try:
            existing_token = None
            
            if token_type == "verification":
                existing_token = await self.get_verification_token_by_email(email);
                if existing_token:
                    await prisma.verificationtoken.delete(
                    where={" id": existing_token.id }
                )
            
            if token_type == "twoFactor":
                existing_token = await self.get_two_factor_token_by_email(email);
                if existing_token:
                    await prisma.twofactortoken.delete(
                        where={" id": existing_token.id }
                    )
                    
        except Exception as e:
            raise Exception(f"Error deleting {token_type} token: {e}")
        
        
        
        
        
    
    async def generate_verification_token(self, email: str) -> Any:
        token = self.generate_random_code()
        expires = datetime.utcnow().replace(tzinfo=pytz.UTC) + timedelta(minutes=10)
        try:
            await self.delete_existing_token(email, 'verification')
            verification_token = await prisma.verificationtoken.create(
                data={
                    "email": email,
                    "token": token,
                    "expires": expires
                }
            )
            
            await self.email_service.send_verification_email(email, token)
            return verification_token
            
        except Exception as error:
            print(f"Error generating verification token for {email}:", error)
            raise Exception('Failed to generate verification token')
        
        
        
        
        
    async def generate_two_factor_token(self, email: str) -> Any:
        token = self.generate_random_code()
        expires = datetime.utcnow().replace(tzinfo=pytz.UTC) + timedelta(minutes=10)
        try:
            await self.delete_existing_token(email, 'twoFactor')
            two_factor_token = await prisma.twofactortoken.create(
                data={
                    "email": email,
                    "token": token,
                    "expires": expires
                }
            )
            
            return two_factor_token
            
        except Exception as error:
            print(f"Error generating two factor token for {email}:", error)
            raise Exception('Failed to generate verification token')